storage-proofs-porep/src/stacked/vanilla/proof.rs

```diff
@@ -554,53 +556,86 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
+                            let selector = GPUSelector::UniqueId(device.device_id().ok_or(
+                                Error::Unclassified(
+                                    "No resource allocation for TreeBuilder".to_string(),
+                                ),
+                            )?);
+                            column_tree_builder =
+                                Some(ColumnTreeBuilder::<ColumnArity, TreeArity>::new(
+                                    Some(BatcherType::CustomGPU(selector)),
+                                    nodes_count,
+                                    max_gpu_column_batch_size,
+                                    max_gpu_tree_batch_size,
+                                )?);
+                        }
 
-                    // Loop until all trees for all configs have been built.
-                    for i in 0..config_count {
-                        loop {
-                            let (columns, is_final): (Vec<GenericArray<Fr, ColumnArity>>, bool) =
-                                builder_rx.recv().expect("failed to recv columns");
+                        let builder = column_tree_builder.as_mut().ok_or(Error::Unclassified(
+                            "failed creating tree builder".to_string(),
+                        ))?;
 
-                            // Just add non-final column batches.
+                        // Loop until all trees for all configs have been built.
+                        if i < config_count {
                             if !is_final {
-                                column_tree_builder
-                                    .add_columns(&columns)
-                                    .expect("failed to add columns");
-                                continue;
-                            };
+                                let (columns, last): (Vec<GenericArray<Fr, ColumnArity>>, bool) =
+                                    builder_rx
+                                        .recv()
+                                        .map_err(|e| Error::Unclassified(e.to_string()))?;
+                                builder.add_columns(&columns)?;
 
-                            // If we get here, this is a final column: build a sub-tree.
-                            let (base_data, tree_data) = column_tree_builder
-                                .add_final_columns(&columns)
-                                .expect("failed to add final columns");
-                            trace!(
-                                "base data len {}, tree data len {}",
-                                base_data.len(),
-                                tree_data.len()
-                            );
+                                is_final = last;
+                            } else {
+                                // If we get here, this is a final column: build a sub-tree.
+                                if let Some((base_data, tree_data)) = builder.build_next()? {
```

Comment:
It looks like you are building a sub-tree without having yet added the final batch of columns. I think you need to add the columns first, then start in on the building.

This brings up the point that with the tentative changes to neptune being used here, the API has been made more brittle. This mistake would not have been possible using the old API. You can probably fix as described above, but I would recommend also modifying `neptune` so it becomes an error to call `build_next()` before the final columns have been added. This would actually add some error-checking (enforcing that trees must be full) that was missing before.

If adding that check causes this code to fail at runtime, you will know it's working. Hopefully then fixing the problem here will stop that error, giving some confidence in the changes on both sides of the API boundary.

Reply:
After looking at this again, I think the actual logic is correct here. However, the variable name (`is_final`) and comments are now inaccurate, leading to the confusion.

Consider that the comment on line 608 suggests possession of a 'final' column. This is a first-class concept inherited from the neptune API (that there be a 'final' batch of columns). However, the final column(s) were actually received and added on the _previous_ invocation of  `call`.

I would change `is_final` to something like `done_adding_columns` to make the meaning very clear. The comment should probably be split into two, one describing what happens in each branch of the conditional.

It may still make sense to add the check to neptune, to rule out the prior possibility, but I don't _think_ the code is 'wrong' after all. That change would make it impossible to create partial trees with implicit zero-padding, but I'm not sure we need to preserve that possibility (though it would technically be a breaking API change).